src/Compilers/Core/Portable/RealParser.cs

```diff
@@ -358,14 +342,24 @@ public static DecimalFloatingPointString FromSource(string source)
                     int firstExponent = i;
                     int lastExponent = i;
                     while (i < source.Length && source[i] >= '0' && source[i] <= '9') lastExponent = ++i;
-                    int exponentMagnitude = int.Parse(source.Substring(firstExponent, lastExponent - firstExponent));
-                    if (exponentSign == '-')
+
+                    int exponentMagnitude = 0;
+
+                    if (int.TryParse(source.Substring(firstExponent, lastExponent - firstExponent), out exponentMagnitude) &&
+                        exponentMagnitude <= MAX_EXP_QUAD)
                     {
-                        exponent -= exponentMagnitude;
+                        if (exponentSign == '-')
+                        {
+                            exponent -= exponentMagnitude;
+                        }
+                        else
+                        {
+                            exponent += exponentMagnitude;
```

Comment:
The parsed exponent plus the implied exponent need to be in an appropriate range, sure. Bit it is not correct to limit the *parsed exponent* in isolation. For example, for the number `100000000000000000000e-20` we have a parsed exponent with a magnitude of `20`, but the final number has an exponent of `0`. I can do the same thing with 16 thousand digits instead of 20 digits, and then your code gives the "wrong" answer of overflow.